File: solutions/csharp/camicia/1/Camicia.cs
using System.Diagnostics.CodeAnalysis;

[SuppressMessage("Design", "CA1050:Declare types in namespaces", Justification = "<Pending>")]

public static class Camicia
{
    public static GameResult SimulateGame(string[] playerA, string[] playerB) 
        => new Game(playerA, playerB)
            .Simulate();

    private class Game(string[] PlayerA, string[] PlayerB)
    {
        public GameResult Simulate()
        {
            GameState state = new(PlayerA, PlayerB);
            GameStatus status;
            do
            {
                status = PlayTrick(state);
            } while (status == GameStatus.Playing);
            
            return new GameResult(status, state.Tricks, state.CardsPlayed);
        }

        private GameStatus PlayTrick(GameState state)
        {
            if (IsLoop(state))
            {
                return GameStatus.Loop;
            }

            state.Tricks++;
            int paymentDue = 0;
            do
            {
                if (DrawCard(state, out Card? card))
                {
                    paymentDue = ResolveInteraction(state, card, paymentDue);
                }
                else
                {
                    AssignWinToOtherPlayerOnIncompletePayment(state);
                }
            } while (!state.Pile.IsEmpty());

            return DetermineGameStatus(state);
        }

        private static int ResolveInteraction(GameState state, Card card, int paymentDue)
        {
            if (paymentDue == 0)
            {
                SwitchTurn(state);
                paymentDue = card.Payment;
            }
            else if (card.Payment == 0)
            {
                paymentDue--;
                if (paymentDue == 0)
                {
                    SwitchTurn(state);
                    ClaimPile(state);
                }
            }
            else
            {
                paymentDue = card.Payment;
                SwitchTurn(state);
            }

            return paymentDue;
        }

        private static void AssignWinToOtherPlayerOnIncompletePayment(GameState state)
        {
            SwitchTurn(state);
            ClaimPile(state);
        }

        private static GameStatus DetermineGameStatus(GameState state) 
            => state.Decks.Any(deck => deck.IsEmpty())
                        ? GameStatus.Finished
                        : GameStatus.Playing;

        private static bool IsLoop(GameState state)
            => !state.History.Add(HashDecks(state));

        private static int HashDecks(GameState state)
            => HashCode.Combine(state.DeckA.EncodePayments(), state.DeckB.EncodePayments());

        private static bool DrawCard(GameState state, out Card card)
        {
            if (!state.Decks[state.Turn].TryDraw(out card))
            {
                return false;
            }

            state.CardsPlayed++;
            state.Pile.Add(card);
            return true;
        }

        private static void SwitchTurn(GameState state)
            => state.Turn = 1 - state.Turn;

        private static void ClaimPile(GameState state)
        {
            state.Decks[state.Turn].AddAll(state.Pile);
            state.Pile.Clear();
        }
    }

    public enum GameStatus
    {
        Playing,
        Finished,
        Loop
    }

    public record GameResult(GameStatus Status, int Tricks, int Cards);
}

internal class GameState
{
    internal GameState(string[] PlayerA, string[] PlayerB)
    {
        DeckA.AddAll(PlayerA);
        DeckB.AddAll(PlayerB);
    }

    internal readonly Deck DeckA = new();
    internal readonly Deck DeckB = new();
    internal Deck[] Decks
        => [DeckA, DeckB];

    internal readonly Deck Pile = new();
    internal int Tricks = 0;
    internal int CardsPlayed = 0;
    internal int Turn = 0;
    internal readonly HashSet<int> History = [];
}

internal record class Card
{
    private static readonly Dictionary<string, int> CardValues = new()
    {
        ["J"] = 1,
        ["Q"] = 2,
        ["K"] = 3,
        ["A"] = 4
    };

    public Card(string card) 
        => Payment = CardValues.TryGetValue(card, out int value)
            ? value
            : 0;
    public int Payment { get; init; }
}

internal class Deck
{
    public Deck() { }
    private Queue<Card> Cards { get; init; } = new();

    public bool IsEmpty() 
        => Cards.Count == 0;

    public bool TryDraw(out Card card) 
        => Cards.TryDequeue(out card!);

    public void Clear() 
        => Cards.Clear();

    public void Add(string card) 
        => Add(new Card(card));

    public void Add(Card card) 
        => Cards.Enqueue(card);

    public void AddAll(IEnumerable<string> cards) 
        => AddAll(cards.Select(card => new Card(card)));

    public void AddAll(IEnumerable<Card> cards)
    {
        foreach (var card in cards)
        {
            Add(card);
        }
    }
    
    protected IEnumerable<Card> AsEnumerable() 
        => Cards;
    
    public void AddAll(Deck deck) 
        => AddAll(deck.AsEnumerable());


    public string EncodePayments() 
        => string.Join(',', Cards.Select(c => c.Payment));
}

File: solutions/java/camicia/1/src/main/java/Camicia.java
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class Camicia {
    static CamiciaResult simulateGame(List<String> playerA, List<String> playerB) {
        return new Camicia()
                .simulate(playerA, playerB);
    }

    public CamiciaResult simulate(List<String> playerA, List<String> playerB) {
        GameState state = new GameState(playerA, playerB);
        GameStatus status;
        do {
            status = this.playTrick(state);
        } while (status == GameStatus.Playing);

        return new CamiciaResult(
                status.toString(),
                state.cardsPlayed,
                state.tricks
        );
    }

    private GameStatus playTrick(GameState state) {
        if (this.isLoop(state)) {
            return GameStatus.Loop;
        }

        state.tricks++;
        int paymentDue = 0;

        do {
            Card card = this.drawCard(state);

            if (card != null) {
                paymentDue = this.resolveInteraction(state, card, paymentDue);
            }
            else {
                this.assignWinToOtherPlayerOnIncompletePayment(state);
            }
        } while (!state.pile.isEmpty());

        return this.determineGameStatus(state);
    }

    private boolean isLoop(GameState state) {
        return !state.history.add(hashDecks(state));
    }

    private int hashDecks(GameState state) {
        int result = 1;
        result = 31 * result + state.deckA.encodePayments().hashCode();
        result = 31 * result + state.deckB.encodePayments().hashCode();
        return result;
    }

    private Card drawCard(GameState state) {
        Optional<Card> optionalCard = state.decks()[state.turn]
                .tryDraw();
        if (optionalCard.isEmpty()) {
            return null;
        }

        state.cardsPlayed++;
        final Card card = optionalCard.get();
        state.pile.add(card);
        return card;
    }

    private int resolveInteraction(GameState state, Card card, int paymentDue) {
        if (paymentDue == 0) {
            this.switchTurn(state);
            paymentDue = card.payment();
        }
        else if (card.payment() == 0) {
            paymentDue--;

            if (paymentDue == 0) {
                this.switchTurn(state);
                this.claimPile(state);
            }
        }
        else {
            paymentDue = card.payment();
            this.switchTurn(state);
        }

        return paymentDue;
    }

    private void assignWinToOtherPlayerOnIncompletePayment(GameState state) {
        this.switchTurn(state);
        this.claimPile(state);
    }

    private GameStatus determineGameStatus(GameState state) {
        return Arrays.stream(state.decks())
                     .anyMatch(Deck::isEmpty)
               ? GameStatus.Finished
               : GameStatus.Playing;
    }

    private void switchTurn(GameState state) {
        state.turn = 1 - state.turn;
    }

    private void claimPile(GameState state) {
        List<Card> temp = StreamSupport
                .stream(state.pile.asEnumerable().spliterator(), false)
                .collect(Collectors.toList());

        state.decks()[state.turn]
                .addAllCards(temp);
        state.pile.clear();
    }
}
